### PR TITLE
Implement forcing of host ip address

### PR DESCRIPTION
**Pull request by 0xfede7c8**

### CODE REVIEW

File: default_options.h [hunk 1]
```diff
@@ -314,4 +314,7 @@ just one connection, and it will stay waiting for the end of the connection, aft
 the program will ends. */
 #define DROPBEAR_SINGLE_SYNCH_CONNECTION 0
 
+/* Enables accept connection from an specific IP */
+#define DROPBEAR_RESTRICT_FIXED_HOST_IP 0
```
Comment (dgutson) [inline default_options.h]: _CLIENT_IP

File: svr-main.c [hunk 2]
```diff
@@ -313,6 +314,15 @@ static void main_noinetd() {
 
 				getaddrstring(&remoteaddr, NULL, &remote_port, 0);
 				dropbear_log(LOG_INFO, "Child connection from %s:%s", remote_host, remote_port);
+
+				#if DROPBEAR_RESTRICT_FIXED_HOST_IP
+				#define MAX_IPV4_LEN 9
+				printf("%s\n%s\n", remote_host, svr_opts.allowed_host_ip_addr);
```
Comment (dgutson) [inline svr-main.c]: remove

File: svr-runopts.c [hunk 3]
```diff
@@ -150,7 +153,10 @@ void svr_getopts(int argc, char ** argv) {
 #if DROPBEAR_SCP_FIXED_FILE_PATH_AND_SIZE
 	svr_opts.allowed_path = NULL;
 	svr_opts.allowed_max_size = NULL;
-#endif 
+#endif
+#if DROPBEAR_RESTRICT_FIXED_HOST_IP
+	svr_opts.allowed_host_ip_addr;
```
Comment (dgutson) [inline svr-runopts.c]: = NULL

File: svr-main.c [hunk 4]
```diff
@@ -313,6 +314,15 @@ static void main_noinetd() {
...
+				if (strcmp(remote_host, svr_opts.allowed_host_ip_addr))
```
Comment (dgutson) [inline svr-main.c]: if P => Q
P: svr_opts.allowed_host_ip_addr != NULL
Q: strcmp(remote_host, svr_opts.allowed_host_ip_addr) == 0

!P || Q

File: svr-runopts.c [hunk 5]
```diff
@@ -459,6 +470,13 @@ void svr_getopts(int argc, char ** argv) {
 		}
 	}
 #endif
+
+#if DROPBEAR_RESTRICT_FIXED_HOST_IP
```
Comment (dgutson) [inline svr-runopts.c]: don't turn this mandatory. Remove this block.

**Review by dgutson — CHANGES_REQUESTED** [review]

File: default_options.h (hunk 1 above)
Comment (0xfede7c8) [reply default_options.h]: Done.

File: svr-main.c (hunk 2 above)
Comment (0xfede7c8) [reply svr-main.c]: Done.

File: svr-runopts.c (hunk 3 above)
Comment (0xfede7c8) [reply svr-runopts.c]: Done.

File: svr-main.c (hunk 4 above)
Comment (0xfede7c8) [reply svr-main.c]: done.

File: svr-runopts.c (hunk 5 above)
Comment (0xfede7c8) [reply svr-runopts.c]: Done.

File: svr-main.c [hunk 6]
```diff
@@ -315,10 +315,8 @@ static void main_noinetd() {
 				getaddrstring(&remoteaddr, NULL, &remote_port, 0);
 				dropbear_log(LOG_INFO, "Child connection from %s:%s", remote_host, remote_port);
 
-				#if DROPBEAR_RESTRICT_FIXED_HOST_IP
-				#define MAX_IPV4_LEN 9
-				printf("%s\n%s\n", remote_host, svr_opts.allowed_host_ip_addr);
-				if (strcmp(remote_host, svr_opts.allowed_host_ip_addr))
+				#if DROPBEAR_RESTRICT_FIXED_CLIENT_IP
+				if ((svr_opts.allowed_client_ip_addr != NULL) && strcmp(remote_host, svr_opts.allowed_client_ip_addr))
```
Comment (dgutson) [inline svr-main.c]: please explicit strcmp != 0

File: svr-main.c [hunk 7]
```diff
@@ -315,10 +315,8 @@ static void main_noinetd() {
 				getaddrstring(&remoteaddr, NULL, &remote_port, 0);
 				dropbear_log(LOG_INFO, "Child connection from %s:%s", remote_host, remote_port);
 
-				#if DROPBEAR_RESTRICT_FIXED_HOST_IP
-				#define MAX_IPV4_LEN 9
-				printf("%s\n%s\n", remote_host, svr_opts.allowed_host_ip_addr);
-				if (strcmp(remote_host, svr_opts.allowed_host_ip_addr))
+				#if DROPBEAR_RESTRICT_FIXED_CLIENT_IP
```
Comment (dgutson) [inline svr-main.c]: Move this to a validate_client_ip(remote_host) function,
provide two versions depending on the macro.

**Review by dgutson — CHANGES_REQUESTED** [review]

File: svr-main.c (hunk 6 above)
Comment (0xfede7c8) [reply svr-main.c]: Done

File: svr-main.c (hunk 7 above)
Comment (0xfede7c8) [reply svr-main.c]: Done

File: svr-main.c [hunk 8]
```diff
@@ -35,6 +35,7 @@ static size_t listensockets(int *sock, size_t sockcount, int *maxfd);
 static void sigchld_handler(int dummy);
 static void sigsegv_handler(int);
 static void sigintterm_handler(int fish);
+static int validate_client_ip(const char*, const char*);
```
Comment (dgutson) [inline svr-main.c]: add arguments

File: svr-main.c [hunk 9]
```diff
@@ -356,6 +356,19 @@ static void main_noinetd() {
 }
 #endif /* NON_INETD_MODE */
 
+#if DROPBEAR_RESTRICT_FIXED_CLIENT_IP
+static int validate_client_ip(const char* remote_host, const char* allowed_client_ip_addr)
+{
+	return (allowed_client_ip_addr != NULL) && (strcmp(remote_host, allowed_client_ip_addr) != 0);
```
Comment (dgutson) [inline svr-main.c]: isn't it a global variable?
The logic is wrong, it should be:
allowed != NULL => they match

File: svr-main.c [hunk 10]
```diff
@@ -356,6 +356,19 @@ static void main_noinetd() {
...
+}
+#else
+static int validate_client_ip(const char* remote_host, const char* allowed_client_ip_addr)
+{
+	(void)remote_host;
+	(void)allowed_client_ip_addr;
+	return 0;
```
Comment (dgutson) [inline svr-main.c]: this turns validation always invalid, breaking the DROPBEAR_RESTRICT_FIXED_CLIENT_IP

**Review by dgutson — CHANGES_REQUESTED** [review]

File: svr-main.c (hunk 9 above)
Comment (dgutson) [reply svr-main.c]: is NULL || they match

File: svr-main.c (hunk 8 above)
Comment (0xfede7c8) [reply svr-main.c]: Done.

File: svr-main.c (hunk 9 above)
Comment (0xfede7c8) [reply svr-main.c]: Done.

File: svr-main.c (hunk 10 above)
Comment (0xfede7c8) [reply svr-main.c]: Didn't because of the inverse logic of validate_client_ip.
Done with new logic.

**Review by dgutson — APPROVED** [review]

**Pull request merged**